Detectors/TRD/simulation/src/Detector.cxx

```diff
@@ -121,48 +134,113 @@ bool Detector::ProcessHits(FairVolume* v)
   if (drRegion && fMC->IsTrackEntering()) {
-    // Create a track reference at the entrance of each
-    // chamber that contains the momentum components of the particle
-    /* Do what AliRoot does here
-      fMC->TrackMomentum(mom);
-      AddTrackReference(gAlice->GetMCApp()->GetCurrentTrackNumber(), AliTrackReference::kTRD);
-    */
     // Create the hits from TR photons if electron/positron is entering the drift volume
-    const int ele = TMath::Abs(fMC->TrackPid()) == 11; // electron PDG code.
+    const bool ele = (TMath::Abs(fMC->TrackPid()) == 11); // electron PDG code.
     if (mTRon && ele) {
-      // Do TR simulation here
-      // Emulate what AliTRDsimTR does here
-      // CreateTRhit(); // See AliTRDv1.cxx
+      CreateTRhit(det);
     }
     trkStat = 1;
   } else if (amRegion && fMC->IsTrackExiting()) {
-    // Create a track reference at the exit of each
-    // chamber that contains the momentum components of the particle
-    /* Do what AliRoot does here
-            fMC->TrackMomentum(mom);
-            AddTrackReference(gAlice->GetMCApp()->GetCurrentTrackNumber(), AliTrackReference::kTRD);
-            trkStat = 2;
-    */
     trkStat = 2;
   }
+
   // Calculate the charge according to GEANT Edep
   // Create a new dEdx hit
-  const double enDep = TMath::Max(fMC->Edep(), 0.0) * 1.0e+9;
+  const double enDep = TMath::Max(fMC->Edep(), 0.0) * 1.0e+9; // Energy in eV
   // Store those hits with enDep bigger than the ionization potential of the gas mixture for in-flight tracks
   // or store hits of tracks that are entering or exiting
   if ((enDep > mWion) || trkStat) {
     double x, y, z;
     fMC->TrackPosition(x, y, z);
-    double time = fMC->TrackTime() * 1e6;
+    double tof = fMC->TrackTime() * 1e6; // The time of flight in micro-seconds
     o2::Data::Stack* stack = (o2::Data::Stack*)fMC->GetStack();
     const int trackID = stack->GetCurrentTrackNumber();
-    addHit(x, y, z, time, enDep, trackID, det);
+    addHit(x, y, z, tof, enDep, trackID, det);
     stack->addHit(GetDetId());
     return true;
   }
   return false;
 }
 
+void Detector::CreateTRhit(int det)
+{
+  //
+  // Creates an electron cluster from a TR photon.
+  // The photon is assumed to be created a the end of the radiator. The
+  // distance after which it deposits its energy takes into account the
+  // absorbtion of the entrance window and of the gas mixture in drift
+  // volume.
+  //
+
+  // Maximum number of TR photons per track
+  constexpr int mMaxNumberOfTRPhotons = 50; // Make this a class member?
+
+  TLorentzVector mom;
+  fMC->TrackMomentum(mom);
+  float pTot = mom.Rho();
+  std::vector<float> photonEnergyContainer;            // energy in keV
+  mTR->createPhotons(11, pTot, photonEnergyContainer); // Create TR photons
+  if (photonEnergyContainer.size() > mMaxNumberOfTRPhotons) {
+    LOG(FATAL) << "Boundary error: nTR = " << photonEnergyContainer.size() << ", mMaxNumberOfTRPhotons = " << mMaxNumberOfTRPhotons;
```

Comment:
this will abort the simulation. Is this a catastrophic condition?

Reply:
It was in AliRoot, I just kept the same severity level. Could @tdietel confirm how catastrophic this could be?

Reply:
I think I disagree. Do you want to the simulation to crash just when you have 1 photon too much? Correct behaviour of the createPhotons functions could be checked in a unit test and we could even get rid of this runtime-if here.

Reply:
Ok I agree. I will reduce the severity to ERROR for now. We can add the unit test later and check if we can remove this runtime-if.